src/lib/parsers/sol-json.c

```diff
@@ -607,40 +607,52 @@ sol_json_calculate_escaped_string_len(const char *str)
 }
 
 SOL_API char *
-sol_json_escape_string(const char *str, char *buf, size_t len)
+sol_json_escape_string(const char *str, struct sol_buffer *buf)
 {
-    char *out = buf;
-    size_t i;
+    char *out, *r_str;
+    size_t i, escaped_len;
+    int r;
 
     SOL_NULL_CHECK(str, NULL);
     SOL_NULL_CHECK(buf, NULL);
 
-    for (i = 0; *str && i < len; str++, i++) {
+    escaped_len = sol_json_calculate_escaped_string_len(str);
+
+    r = sol_buffer_expand(buf, escaped_len);
```

Comment:
you should only expand what is missing. Eventually there is enough capacity to fit the escaped string and nothing needs to be expanded.